rclcpp/src/rclcpp/node_interfaces/node_base.cpp

```diff
@@ -0,0 +1,195 @@
+// Copyright 2016 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "rclcpp/node_interfaces/node_base.hpp"
+
+#include "rclcpp/exceptions.hpp"
+
+using rclcpp::exceptions::throw_from_rcl_error;
+
+using namespace rclcpp::node_interfaces;
+
+NodeBase::NodeBase(
+  const std::string & node_name,
+  rclcpp::context::Context::SharedPtr context)
+: name_(node_name),
+  context_(context),
+  node_handle_(nullptr),
+  default_callback_group_(nullptr),
+  associated_with_executor_(false),
+  notify_guard_condition_is_valid_(false)
+{
+  // Setup the guard condition that is notified when changes occur in the graph.
+  rcl_guard_condition_options_t guard_condition_options = rcl_guard_condition_get_default_options();
+  rcl_ret_t ret = rcl_guard_condition_init(&notify_guard_condition_, guard_condition_options);
+  if (ret != RCL_RET_OK) {
+    throw_from_rcl_error(ret, "failed to create interrupt guard condition");
+  }
+
+  // Setup a safe exit lamda to clean up the guard condition in case of an error here.
+  auto finalize_notify_guard_condition = [this]() {
+    // Finalize the interrupt guard condition.
+    if (rcl_guard_condition_fini(&notify_guard_condition_) != RCL_RET_OK) {
+      fprintf(stderr,
+        "[rclcpp::error] failed to destroy guard condition: %s\n", rcl_get_error_string_safe());
+    }
+  };
+
+  // Determine the domain id based on the options and the ROS_DOMAIN_ID env variable.
+  size_t domain_id = 0;
+  char * ros_domain_id = nullptr;
+  const char * env_var = "ROS_DOMAIN_ID";
+#ifndef _WIN32
+  ros_domain_id = getenv(env_var);
+#else
+  size_t ros_domain_id_size;
+  _dupenv_s(&ros_domain_id, &ros_domain_id_size, env_var);
+#endif
+  if (ros_domain_id) {
+    uint32_t number = strtoul(ros_domain_id, NULL, 0);
+    if (number == (std::numeric_limits<uint32_t>::max)()) {
+      // Finalize the interrupt guard condition.
+      finalize_notify_guard_condition();
+
+      throw std::runtime_error("failed to interpret ROS_DOMAIN_ID as integral number");
```

Comment:
While this code block has only been moved this line leaks the memory of `ros_domain_id` on Windows.